### PR TITLE
[OLD] Allow returning of the state

### PR DESCRIPTION
**Pull request by trbromley**
This PR allows users to return the quantum state using a new `state()` function:

```python
import pennylane as qml

dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def f(x):
    qml.RX(x, wires=0)
    qml.RY(0.4, wires=1)
    qml.CNOT(wires=[0, 1])
    return qml.state()

f(0.3)
```

**Not currently implemented:**

- Support for differentiating the state
- Support for returning the state in CV devices
- Support for returning the state in versions of torch before 1.6.0 (since complex numbers are not supported)

**Questions:**

- The state should be returned on all wires, hence the new `state()` return type loads the observable `op = qml.Identity(wires=AllWires, do_queue=False)`. Is using `AllWires` acceptable? I can see two side effects:

  - Good side effect: users can return the state (of all wires) at the same time as other return types such as `expval`. This seems ok, since anyway the state is available on simulator devices.
  - Bad side effect: the circuit drawer places the word "State" after only the first wire and not all of them. There may also be incompatibilities when combined with other return functions, e.g., if one returns `qml.expval(qml.PauliZ(0))`, there is an `IndexError`.

- The `metric_tensor()` method of `QubitQNode` was not updated, in line with not offering support for differentiation of the state.
- The state is still analytic when the device is in `non_analytic` mode. I think this makes sense, but just wanted to make a note of it.

### CODE REVIEW

File: pennylane/measure.py [hunk 1]
```diff
@@ -199,3 +199,17 @@ def circuit():
...
+
+
+from pennylane.operation import AllWires, State
+
+
+def state():
+    r"""TODO
+    """
+    op = qml.Identity(wires=AllWires, do_queue=False)
+    op.return_type = State
```
Comment (josh146) [inline pennylane/measure.py:211]: I would monkeypatch the gradient recipe here

File: pennylane/_qubit_device.py [hunk 2]
```diff
@@ -280,6 +280,9 @@ def statistics(self, observables):
             elif obs.return_type is Probability:
                 results.append(self.probability(wires=obs.wires))
 
+            elif obs.return_type is State:
+                results.append(self.state)
```
Comment (josh146) [inline pennylane/_qubit_device.py:284]: don't forget to check if `self.state` is defined 😆 or better yet, try/except

Comment (trbromley) [reply pennylane/_qubit_device.py:284]: Thanks, done!

File: pennylane/_qubit_device.py [hunk 3]
```diff
@@ -280,6 +280,17 @@ def statistics(self, observables):
...
                 results.append(self.probability(wires=obs.wires))
 
+            elif obs.return_type is State:
+                try:
+                    state = self.state
+                except NotImplementedError:
+                    state = None
+
+                if state is None:
+                    raise ValueError("The state is not available in the current device")
```
Comment (josh146) [inline pennylane/_qubit_device.py:290]: I was going to suggest combining it into a single
```python
except NotImplementedError:
    raise ValueError("The state is not available in the current device")
```

But I suppose this is better, since some devices might return `state=None`?

File: pennylane/qnodes/jacobian.py [hunk 4]
```diff
@@ -323,7 +327,11 @@ def inds_using(m):
                 if variances_required:
                     grad[:, i] = self._pd_analytic_var(k, flat_args, kwargs, **options)
                 else:
-                    grad[:, i] = self._pd_analytic(k, flat_args, kwargs, **options)
+                    if state_required:
+                        factor = 2 * np.sin(self.h) / (4 * np.sin(self.h / 2))
+                    else:
+                        factor = 1
+                    grad[:, i] = factor * self._pd_analytic(k, flat_args, kwargs, **options)
```
Comment (josh146) [inline pennylane/qnodes/jacobian.py:334]: Perhaps, to make it a bit cleaner?
```suggestion
                    grad[:, i] = self._pd_analytic(k, flat_args, kwargs, **options)

                    if state_required:
                        grad[:, i] *= 2 * np.sin(self.h) / (4 * np.sin(self.h / 2))
```

Comment (josh146) [reply pennylane/qnodes/jacobian.py:334]: Just realised that `2 * np.sin(self.h) / (4 * np.sin(self.h / 2))` simplifies down to just `np.cos(self.h / 2)` 😆

File: pennylane/_qubit_device.py (hunk 3 above)
Comment (trbromley) [reply pennylane/_qubit_device.py:293]: Yes, I originally thought that but found out the ones that return `None`

File: pennylane/interfaces/torch.py [hunk 5]
```diff
@@ -199,6 +204,19 @@ def forward(ctx, input_kwargs, *input_):
...
+                    for op in qnode.ops
+                ]
+            )
+
+            if returns_state and not MIN_VERSION_FOR_STATE:
+                raise ImportError(
+                    "Version 1.6.0 or above of PyTorch must be installed to return the quantum "
+                    "state using the torch interface"
+                )
+
```
Comment (trbromley) [inline pennylane/interfaces/torch.py:219]: I know we discussed not having an explicit check for the torch version and mentioning it in the docs - but I feel that given the recentness of the only useable release for complex numbers, an explicit check might help for now.

File: tests/qnodes/test_qnode_base.py [hunk 6]
```diff
@@ -681,6 +681,21 @@ def circuit(x, y=0.2, *args, z=0.3):
...
+
+        def circuit():
+            return qml.state()
+
+        node = BaseQNode(circuit, dev)
+
+        with pytest.raises(NotImplementedError, match="Returning the quantum state with"):
+            node()
+
+
```
Comment (trbromley) [inline tests/qnodes/test_qnode_base.py:697]:
```suggestion
```

File: tests/test_qubit_device.py [hunk 7]
```diff
@@ -66,8 +66,9 @@ def mock_qubit_device_extract_stats(monkeypatch):
         m.setattr(QubitDevice, "expval", lambda self, x: 0)
         m.setattr(QubitDevice, "var", lambda self, x: 0)
         m.setattr(QubitDevice, "sample", lambda self, x: 0)
+        m.setattr(QubitDevice, "state", 0)
         m.setattr(
-            QubitDevice, "probability", lambda self, wires=None: 0 if wires is None else wires
+            QubitDevice, "probability", lambda self, wires=None: 0
```
Comment (trbromley) [inline tests/test_qubit_device.py:71]: I wasn't sure why we originally had ` if wires is None else wires`? But the test wouldn't work with it (possibly due to the wires refactor)

File: tests/test_qubit_device.py [hunk 8]
```diff
@@ -326,8 +327,26 @@ class SomeObservable(qml.operation.Observable):
...
+
+        with monkeypatch.context() as m:
+            dev = mock_qubit_device_extract_stats()
+            delattr(dev.__class__, "state")
+            with pytest.raises(AttributeError, match="The state is not available in the current"):
+                dev.statistics([obs])
+
     @pytest.mark.parametrize("returntype", [None])
-    def test_results_created(self, mock_qubit_device_extract_stats, monkeypatch, returntype):
+    def test_results_created_empty(self, mock_qubit_device_extract_stats, monkeypatch, returntype):
```
Comment (trbromley) [inline tests/test_qubit_device.py:349]: Note that previously the test name clashed with `test_results_created` above

File: pennylane/measure.py [hunk 9]
```diff
@@ -199,3 +208,42 @@ def circuit():
...
+    >>> circuit()
+    array([0.70710678+0.j, 0.70710678+0.j, 0.        +0.j, 0.        +0.j])
+
+    The returned array is in lexicographic order, so corresponds
+    to a :math:`1/\sqrt{2}` amplitude in both :math:`|00\rangle`
+    and :math:`|01\rangle`.
+
+    .. note::
+
+        The ``state()`` function does not currently support differentiation of the returned state.
```
Comment (josh146) [inline pennylane/measure.py:242]: This won't stop people from trying to differentiate it thought 😆

File: pennylane/qnodes/base.py [hunk 10]
```diff
@@ -664,7 +664,14 @@ def _check_circuit(self, res):
                     num_basis_states = getattr(self.device, "cutoff", 10)
 
                 self.output_dim = num_basis_states ** len(res.wires)
+            elif res.return_type is ObservableReturnTypes.State:
+                if self.model == "cv":
+                    raise NotImplementedError(
+                        "Returning the quantum state with a CV device is " "not yet supported"
```
Comment (josh146) [inline pennylane/qnodes/base.py:670]:
```suggestion
                        "Returning the quantum state with a CV device is not yet supported"
```

File: tests/test_measure.py [hunk 11]
```diff
@@ -241,3 +241,149 @@ def circuit():
...
+
+    def test_combination(self):
+        """Test that the state can be output in combination with other return types."""
+        dev = qml.device("default.qubit", wires=4)
+
+        @qml.qnode(dev)
+        def func():
+            for i in range(4):
+                qml.Hadamard(i)
+            return qml.state(), qml.expval(qml.PauliZ(0)), qml.var(qml.PauliX(1))
```
Comment (josh146) [inline tests/test_measure.py:320]: Note `func.draw()` will fail, since we have two measurements on wire 0 🤔 Likely other things might break as well

File: tests/test_measure.py [hunk 12]
```diff
@@ -241,3 +241,149 @@ def circuit():
...
+        state_expected = 0.25 * tf.ones(16)
+        state = func()
+
+        assert isinstance(state, tf.Tensor)
+        assert state.dtype == tf.complex128
+        assert np.allclose(state_expected, state.numpy())
+
+    def test_interface_torch(self):
+        """Test that the state correctly outputs in the torch interface"""
+        torch = pytest.importorskip("torch", minversion="1.6")
```
Comment (josh146) [inline tests/test_measure.py:352]: nice!

File: pennylane/interfaces/torch.py (hunk 5 above)
Comment (josh146) [reply pennylane/interfaces/torch.py:219]: No this is great! In that comment I meant instead not having a specific check for deprecating support for old minor versions.

E.g., if we drop support for Torch 1.3 (and require Torch 1.4+ in the future), we should have checks in the code raising exceptions for Torch <= 1.3.

**Comment by josh146** [comment]
By the way: differentiation of the state via `default.qubit.tf` works really well!

```python
import pennylane as qml
import tensorflow as tf

dev = qml.device("default.qubit.tf", wires=3)

@qml.qnode(dev, interface="tf", diff_method="backprop")
def qnode(p):
  qml.RZ(p[0], wires=0)
  qml.CNOT(wires=[0, 1])
  qml.RY(p[1], wires=1)
  qml.CNOT(wires=[0, 2])
  return qml.state()

params = tf.Variable([0.2, 0.1])

with tf.GradientTape() as tape:
    res = qnode(params)
    loss = tf.math.real(res[0, 0])

print("Loss:", loss)

grad = tape.gradient(loss, params)
print("Gradient:", grad)
```

gives
```
Loss: tf.Tensor(0.9937606689798755, shape=(), dtype=float64)
Gradient: tf.Tensor([-0.04985433 -0.02486474], shape=(2,), dtype=float32)
```

One subtlety however is that the returned state has shape `(1, 2**num_wires)`, not `(2**num_wires,)`. This is probably a result of `np.squeeze` not being called by the PassthruQNode.

On the other hand, I can't seem to get `default.qubit.autograd` gradient to work:

```python
import pennylane as qml
from pennylane import numpy as np

dev = qml.device("default.qubit.autograd", wires=3)

@qml.qnode(dev, interface="autograd", diff_method="backprop")
def qnode(p):
  qml.RZ(p[0], wires=0)
  qml.CNOT(wires=[0, 1])
  qml.RY(p[1], wires=1)
  qml.CNOT(wires=[0, 2])
  return qml.state()

params = np.array([0.2, 0.1], requires_grad=True)

def loss_fn(param):
    res = qnode(params)
    return np.real(res[0, 0])

print("Loss:", loss_fn(params))

grad_fn = qml.grad(loss_fn)
print("Gradient:", grad_fn(params))
```

gives
```
Loss: 0.9937606691655043
/home/josh/miniconda3/envs/37/lib/python3.7/site-packages/autograd/tracer.py:14: UserWarning: Output seems independent of input.
  warnings.warn("Output seems independent of input.")
Gradient: (array([0., 0.]),)
```

🙁

File: pennylane/qnodes/jacobian.py (hunk 4 above)
Comment (trbromley) [reply pennylane/qnodes/jacobian.py]: Resolving - something we can come back to when we add support for differentiating the state with the parameter shift rule.

File: pennylane/measure.py [hunk 13]
```diff
@@ -199,3 +208,42 @@ def circuit():
...
+    """
+    op = qml.Identity(wires=AllWires, do_queue=False)
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/measure.py:244]: Thanks @josh146! I think this is the main remaining thing to sort out 🤔 It did feel a bit weird to use `AllWires` but strangely turned out ok, although I'm worried it's a bit hacky and maybe not such a good idea long term.

> Perhaps 'state' shouldn't count as doing any measurement?

How would that work, as in somehow instantiating `qml.identity()` without any wires?

The "by the book" alternative approach I could think of would be:
- Add a `wires` argument to `qml.sample()`
- Check somewhere if `wires` is not equal to all of the wires, raise an error

The disadvantages of this are:
- Users have to explicitly put in a `wires` argument, even though `qml.sample()` doesn't really need it to function, more due to our internal logic. As a counterargument, maybe explicit is clearer!
- Users would only ever be able to return the state and cannot combine with other return types, e.g., `return qml.state(wires=range(num_wires)), qml.expval(qml.PauliZ(0))` would not be allowed. As a counterargument, maybe this is the less risky approach anyway!

[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/measure.py:244]: > How would that work, as in somehow instantiating `qml.identity()` without any wires?

We would have to work it out, there currently is no way of specifying a return without explicitly measuring wires (but perhaps something we should consider).

> The "by the book" alternative approach I could think of would be:

Yep, I think the by-the-book approach would be `qml.state(wires=[0,1,2])` etc.

File: pennylane/_qubit_device.py [hunk 14]
```diff
@@ -283,6 +283,17 @@ def statistics(self, observables):
             elif obs.return_type is Probability:
                 results.append(self.probability(wires=obs.wires))
 
+            elif obs.return_type is State:
+                try:
+                    state = self.state
+                except AttributeError:
+                    state = None
+
+                if state is None:
```
Comment (co9olguy) [inline pennylane/_qubit_device.py:292]: Though I'm not really well-versed in this things, I *think* the pythonic way to do this (essentially intercepting an exception and changing the message) is to `raise ... from ...`

Comment (co9olguy) [reply pennylane/_qubit_device.py:292]: Also, this is something that a device can explicitly register whether it supports (in the device._capabilities` dictionary

Comment (josh146) [reply pennylane/_qubit_device.py:292]: Good point. I still tend to prefer EAFP in cases like this (try and get the state, pass if there is an error). I worry about maintenance with the `device._capabilities` dictionary --- with EAFP, we can 'activate' new plugin features directly from core. Especially since the (current) device API explicitly raises a `NotImplementedError` if `state()` is not defined.

More complicated interconnection (such as passthru) should definitely use the capabilities attribute.

**Comment by co9olguy** [comment]
Note: this PR has been put on hold until the next release, as we sort out better some of the conceptual/implementation questions that have come up

File: .github/CHANGELOG.md [hunk 15]
```diff
@@ -2,6 +2,32 @@
 
 <h3>New features since last release</h3>
 
+* The quantum state of a QNode can now be returned using the ``state()`` return function.
```
[the remark this answers is not in the document]
Comment (trbromley) [reply .github/CHANGELOG.md:5]: Done 1275916a8ba332ff44303efb4becf37c4f628afa (in the new branch)

File: pennylane/_qubit_device.py (hunk 14 above)
Comment (trbromley) [reply pennylane/_qubit_device.py:292]: Linked conversation https://github.com/PennyLaneAI/pennylane/pull/789/files#r486622239.

File: pennylane/measure.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/measure.py:244]: I've resolved this in the quantum-tape-return-state branch by simply requiring the wires to be supplied by the user.

**Comment by trbromley** [comment]
I am now closing this PR as it is replaced with https://github.com/PennyLaneAI/pennylane/pull/789.
